library/fsio.c

```diff
@@ -0,0 +1,417 @@
+/*
+ *  Serialization based implementation of filesystem IO functions.
+ *
+ *  Copyright (C) 2017, ARM Limited, All Rights Reserved
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License"); you may
+ *  not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ *  This file is part of mbed TLS (https://tls.mbed.org)
+ */
+
+#if !defined(MBEDTLS_CONFIG_FILE)
+#include "mbedtls/config.h"
+#else
+#include MBEDTLS_CONFIG_FILE
+#endif
+
+#if defined(MBEDTLS_FS_IO)
+
+#include <string.h>
+#include <assert.h>
+#include "mbedtls/fsio.h"
+
+#if !defined(MBEDTLS_FS_IO_ALT)
+
+#if !defined(MBEDTLS_PLATFORM_NO_STD_FUNCTIONS)
+/**
+ * \brief          Open file and disable buffering.
+ *
+ * \param path     File path
+ * \param mode     Open mode
+ *
+ * \return         Pointer to mbedtls_file_t on success or NULL on failure.
+ */
+mbedtls_file_t * mbedtls_fopen( const char *path, const char *mode )
+{
+    mbedtls_file_t * f = fopen( path, mode );
+
+    if( f && ( setvbuf( f, NULL, _IONBF, 0 ) != 0 ) )
+    {
+        fclose( f );
+        f = NULL;
+    }
+
+    return( f );
+}
+
+/**
+ * \brief           Rework of readdir to adapt to abstraction. This does not
+ *                  return struct dirent pointer. Hence no need to worry
+ *                  pointer ownership and cleanup.
+ *
+ * \param dir       Pointer to directory handle.
+ * \param file_name Out buffer for directory entry name.
+ *                  Upto 255 character long name can be returned.
+ * \param size      Out buffer length.
+ * \param type      Out pointer entry type.
+ *
+ * \return          0 for success. Non zero for failure.
+ */
+int mbedtls_readdir( mbedtls_dir_t * dir, char * file_name, uint32_t size,  uint32_t * type )
+{
+    int status = -1;
+    struct dirent * entry;
+
+    if ( ( entry = readdir( dir ) ) != NULL )
+    {
+        strncpy( file_name, entry->d_name, size );
+        if ( file_name[size - 1] == '\0' ) /* Check if buffer was enough */
+        {
+            *type = entry->d_type;
+            status = 0;
+        }
+        else
+        {
+            file_name[size - 1] = 0;
+        }
+    }
+
+    return( status );
+}
+
+#endif /* !MBEDTLS_PLATFORM_NO_STD_FUNCTIONS */
+
+#else /* !MBEDTLS_FS_IO_ALT */
+
+#if defined(MBEDTLS_SERIALIZE_C)
+
+#include "mbedtls/serialize.h"
+
+
+
+#define INT_TO_FILE_PTR( x ) ( (mbedtls_file_t *)(uintptr_t)x )
+#define INT_TO_DIR_PTR( x ) ( (mbedtls_dir_t *)(uintptr_t)x )
+#define PTR_TO_INT( x ) ( (int32_t)(uintptr_t)x )
+
```

Comment:
What's the reason to convert to/from unsigned integers first? There is also a signed integer type `intptr_t` for conversion to/from `void*`.

The assumptions on the type of `x` should be documented here. If the macro is supposed to be used for a single type only, I'd prefer to use `static inline` functions instead. The compiler will then warn if we use them with unexpected types.

Reply:
Removed these and their use.